src/runner.c

```diff
@@ -0,0 +1,134 @@
+/*
+** © 2018 by Philipp Dunkel, Ben Noordhuis, Elan Shankar
+** Licensed under MIT License.
+*/
+
+#include "runner.h"
+#include <pthread.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <strings.h>
+#include "CoreFoundation/CoreFoundation.h"
+#define NAPI_EXPERIMENTAL
+#include <node_api.h>
+
+typedef struct fse_event {
+  char path[PATH_MAX];
+  size_t pathlen;
+  UInt64 id;
+  UInt32 flags;
+  struct fse_event *next;
+} fse_event_t;
+
+void handleEvents(
+  ConstFSEventStreamRef stream,
+  void *data,
+  size_t numEvents,
+  void *eventPaths,
+  const FSEventStreamEventFlags eventFlags[],
+  const FSEventStreamEventId eventIds[]
+) {
+  size_t idx;
+  fse_t *instance = (fse_t *)data;
+
+  uv_mutex_lock(&instance->mutex);
+  for (idx=0; idx < numEvents; idx++) {
+    fse_event_t *event = (fse_event_t*)malloc(sizeof(fse_event_t));
+    CFStringRef path = (CFStringRef)CFArrayGetValueAtIndex((CFArrayRef)eventPaths, idx);
+    event->pathlen = CFStringGetLength(path);
+    strncpy(event->path, CFStringGetCStringPtr(path, kCFStringEncodingUTF8), PATH_MAX);
+    event->id = eventIds[idx];
+    event->flags = eventFlags[idx];
+    event->next = NULL;
+    if (instance->head) {
+      ((fse_event_t*)instance->head)->next = event;
+      instance->head = event;
+    } else {
+      instance->head = event;
+      instance->tail = event;
+    }
```

Comment:
Since you're on macos, you could use TAILQ_INSERT_AFTER from `<sys/queue.h>`. Or steal [libuv's queue.h](https://github.com/libuv/libuv/blob/v1.x/src/queue.h), easier to use.

Reply:
Either version looked insanely complex for a simple linked list.

TODO: Leaen cool macros

I'll leave it as is for now, since the cursory read of these I did, didn't lead to me even ahving a clue how to use them.

Reply:
In this case inserting a single event is O(1) and takinging off a single event is also O(1).
 I'll leave it as is and put reading up on the TAILQ_XXX macros on my TODO-list.